### ISSUE TITLE
Inconsistent array concatenation behaviour with respect to NULL values

### ISSUE THREAD

**Issue by lukaseder** [duckdb/duckdb#11781]
### What happens?

Depending on the type of a `NULL` value, the concatenation behaviour is inconsistent for arrays.

### To Reproduce

```sql
select array[1] || null, array[1] || cast(null as int array)
```

This produces:

```
|((ARRAY[1]) || NULL)|((ARRAY[1]) || CAST(NULL AS INTEGER[]))|
|--------------------|---------------------------------------|
|                    |[1]                                    |
```

PostgreSQL prefers the latter (concatenating `NULL` arrays has no effect), but I think that it would be more consistent with other operations to prefer the former, where concatenating `NULL` produces `NULL`. Other dialects, like H2, HSQLDB, etc. implement the former.

### OS:

Microsoft Windows [Version 10.0.22631.3447]

### DuckDB Version:

0.10.2

### DuckDB Client:

JDBC

### Full Name:

Lukas Eder

### Affiliation:

Data Geekery

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Not applicable - the reproduction does not require a data set

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by soerenwolfers** [comment]
I know this is not a vote, but _please_ resolve this to return `NULL`.

Returning `NULL` when operands are `NULL` (except for special null handling functions like `COALESCE` and `IFNULL`) is one of the golden properties of SQL without which it is so much harder to predict how things behave.

Having `[1] || NULL = [1]` is akin to having `1 + NULL = 1`.

**Comment by maiadegraaf** [comment]
Hi, thanks for your issue.

We currently do follow postgres here:
```sql
postgres=# select concat('hi' , NULL);
 concat 
--------
 hi
(1 row)

postgres=# select 'hi' || NULL;
 ?column? 
----------
 
(1 row)
```

DuckDB:
```sql
D select concat('hi', NULL);
┌────────────────────┐
│ concat('hi', NULL) │
│      varchar       │
├────────────────────┤
│ hi                 │
└────────────────────┘

D select 'hi' || NULL;
┌────────────────┐
│ ('hi' || NULL) │
│    varchar     │
├────────────────┤
│                │
└────────────────┘
```

However, in your example, the first instance is a bug.  The optimizer prefers the `VARCHAR` concat over the `LIST` concat.
```cpp
D select concat([1], NULL);
┌──────────────────────────────────┐
│ concat(main.list_value(1), NULL) │
│             varchar              │
├──────────────────────────────────┤
│ [1]                              │
└──────────────────────────────────┘
```

This is being solved in #12317.

**Comment by lukaseder** [comment]
My point here is that the behaviour is inconsistent with respect to whether the `NULL` value is cast to a type or not, which shouldn't matter in my opinion.

When running:

```sql
select array[1] || null, array[1] || cast(null as int array)
```

**DuckDB:** (inconsistent)

```
|((ARRAY[1]) || NULL)|((ARRAY[1]) || CAST(NULL AS INTEGER[]))|
|--------------------|---------------------------------------|
|                    |[1]                                    |
```

**PostgreSQL:** (consistent)

```
|?column?|?column?|
|--------|--------|
|{1}     |{1}     |
```

**H2:** (consistent)

```
|CAST(NULL AS INTEGER ARRAY)|CAST(NULL AS INTEGER ARRAY)|
|---------------------------|---------------------------|
|                           |                           |
```

**Comment by soerenwolfers** [comment]
@maiadegraaf the concat operator is different, and supposed to be different, from the concat function, in that the former returns NULL whenever one of the inputs is NULL while the latter years NULL as identity, right? Doesn't that mean that the second example in the OP is wrong, not the first? And it's wrong not because it picks a string variant over an array variant but because it picks a concat function over a concat operator?

**Comment by taniabogatsch** [comment]
> My point here is that the behaviour is inconsistent with respect to whether the NULL value is cast to a type or not, which shouldn't matter in my opinion.

The inconsistent behavior between a typed `NULL` and an untyped `NULL` is fixed in https://github.com/duckdb/duckdb/pull/12317 as part of the bug mentioned by @maiadegraaf. Have a look at the newly introduced tests [here](https://github.com/duckdb/duckdb/pull/12317/files) (`test/sql/function/string/test_concat_binding.test`).

```sql
query I
select array[1] || cast(NULL as int array);
----
NULL
```

@soerenwolfers, you raised the remaining question of this issue. Currently, in https://github.com/duckdb/duckdb/pull/12317, we follow the same behavior as described in your comment. While we usually aim for compatibility with Postgres, they seem to be inconsistent about `ARRAY` concatenation themselves. Looking at the docs [here](https://www.postgresql.org/docs/current/functions-array.html), it says the following: **Concatenating a null or empty array is a no-op**.

> anycompatiblearray || anycompatiblearray → anycompatiblearray
Concatenates the two arrays. Concatenating a null or empty array is a no-op; otherwise the arrays must have the same number of dimensions (as illustrated by the first example) or differ in number of dimensions by one (as illustrated by the second). If the arrays are not of identical element types, they will be coerced to a common type (see [Section 10.5](https://www.postgresql.org/docs/current/typeconv-union-case.html)).
ARRAY[1,2,3] || ARRAY[4,5,6,7] → {1,2,3,4,5,6,7}
ARRAY[1,2,3] || ARRAY[[4,5,6],[7,8,9.9]] → {{1,2,3},{4,5,6},{7,8,9.9}}

```sql
postgres=# select array[1] || null, array[1] || cast(null as int array);
 ?column? | ?column? 
----------+----------
 {1}      | {1}
(1 row)
```

Thus, we decided to follow `Returning NULL when operands are NULL`.

**Comment by lukaseder** [comment]
> The inconsistent behavior between a typed `NULL` and an untyped `NULL` is fixed in #12317 as part of the bug mentioned by @maiadegraaf.

Ah, I see, thanks for the clarification. I must have overlooked that

**Comment by taniabogatsch** [comment]
No worries! That's why I wrote the clarification. 😄
